### PR TITLE
feat: integration Kimi-Linear-48B model

### LINKED ISSUE #945: [Task]Implement Kimi-Linear model for Kimi-Linear-48B serving
## Goal
Implement the `kimi.py` model framework file under `python/sgl_jax/srt/models/`, assembling MLA Attention, KDA Linear Attention, MoE, MLP, and other components. Handle the interface differences between existing modules in this repository and the official HuggingFace `modeling_kimi.py` (moonshotai/Kimi-Linear-48B-A3B-Instruct), enabling SGL-JAX to load and perform inference on this model.

**In scope:** Model framework file, DecoderLayer assembly logic, weight mapping (WeightMapping), model registration (EntryClass), component interface adaptation.
**Out of scope:** KDA Pallas kernel implementation from scratch, new MoE routing kernel implementation, performance optimization.

**Reference implementation:** https://huggingface.co/moonshotai/Kimi-Linear-48B-A3B-Instruct/blob/main/modeling_kimi.py

## Acceptance Criteria
1. Model file `kimi_linear.py` exists under `python/sgl_jax/srt/models/`, containing `KimiDecoderLayer`, `KimiModel`, `KimiForCausalLM`
2. `EntryClass` registration is successful, `ModelRegistry` can resolve the Kimi architecture
3. `WeightMapping` correctly maps HuggingFace weights to SGL-JAX parameters
4. Hybrid layer logic is correct: MLA and KDA layers are assembled alternately based on config
5. MoE and dense MLP layers are correctly assigned based on config

### PR DESCRIPTION
**Pull request by zhengkezhou1**
## Motivation

Resolve #945 

Add support for [Kimi-Linear-48B-A3B-Instruct](https://huggingface.co/moonshotai/Kimi-Linear-48B-A3B-Instruct), a hybrid architecture model (48B total / ~3B active params) combining MLA + KDA attention with MoE sparse FFN.

## Modifications

- `KimiLinearConfig`: model config with hybrid attention dispatch (`is_kda_layer`, `is_linear_attn`)
- `KimiDecoderLayer`: single-class if/else pattern — MLA layers use existing `MLAAttention` directly, KDA layers use a `KimiDeltaAttention` placeholder (not yet implemented)
- MoE components (`GateLogit`, `TopK`, `EPMoE`, shared experts) are flat on the decoder layer, following the DeepSeek V3 pattern
- Full `WeightMapping` for MLA + dense MLP + MoE layers (KDA mappings deferred until KDA module is implemented)

## Checklist

- [ ] Please use English, otherwise it will be closed.
- [ ] The purpose of the PR, or link existing issues this PR will resolve.
- [ ] The test plan, such as providing test command.
- [ ] (Optional) The necessary documentation update.

### CODE REVIEW

File: python/sgl_jax/srt/models/kimi_linear.py [hunk 1]
```diff
@@ -0,0 +1,677 @@
...
+from sgl_jax.srt.layers.logits_processor import LogitsMetadata, LogitsProcessor
+from sgl_jax.srt.layers.moe import EPMoE, create_moe_weights_mapping
+from sgl_jax.srt.mem_cache.memory_pool import KVCache
+from sgl_jax.srt.model_executor.forward_batch_info import ForwardBatch
+from sgl_jax.srt.utils.weight_utils import WeightLoader, WeightMapping
+
+logger = logging.getLogger(__name__)
+
+
+class KimiLinearConfig(PretrainedConfig):
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/models/kimi_linear.py]: Place the config to `srt/configs/kimi_linear.py`

File: python/sgl_jax/srt/models/kimi_linear.py [hunk 2]
```diff
@@ -0,0 +1,677 @@
...
+                qk_nope_head_dim=config.qk_nope_head_dim,
+                qk_rope_head_dim=config.qk_rope_head_dim,
+                v_head_dim=config.v_head_dim,
+                mesh=mesh,
+                layer_id=layer_idx,
+                rope_theta=config.rope_theta,
+                rope_scaling=config.rope_scaling,
+                rope_interleave=True,
+                dtype=dtype,
+            )
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/models/kimi_linear.py]: Use `from sgl_jax.srt.models.deepseek_v3 import DeepseekV3Attention as KimiMLAAttention`

File: python/sgl_jax/srt/models/kimi_linear.py (hunk 1 above)
Comment (zhengkezhou1) [reply python/sgl_jax/srt/models/kimi_linear.py]: done. moved to `srt/configs/kimi_linear.py`

File: python/sgl_jax/srt/models/kimi_linear.py (hunk 2 above)
Comment (zhengkezhou1) [reply python/sgl_jax/srt/models/kimi_linear.py]: done. using DeepseekV3Attention

**Comment by aolemila** [comment]
Please rebase epic/support_kimi_linear, there are conflicts. @zhengkezhou1

File: python/sgl_jax/srt/models/kimi_linear.py [hunk 3]
```diff
@@ -0,0 +1,543 @@
...
+                kv_lora_rank=config.kv_lora_rank,
+                qk_nope_head_dim=config.qk_nope_head_dim,
+                qk_rope_head_dim=config.qk_rope_head_dim,
+                v_head_dim=config.v_head_dim,
+                mesh=mesh,
+                layer_id=layer_idx,
+                rope_theta=config.rope_theta,
+                rope_scaling=config.rope_scaling,
+                rope_interleave=True,
+                dtype=dtype,
```
Comment (aolemila) [inline python/sgl_jax/srt/models/kimi_linear.py:310]: @zhengkezhou1 Current we have supported absorbed_mla, so you can refer to initilization of `DeepseekV3Attention` and add `use_absorbed=getattr(config, "use_absorbed_mla", True)` when initilizing `KimiMLAAttention`.

File: python/sgl_jax/srt/models/kimi_linear.py [hunk 4]
```diff
@@ -0,0 +1,543 @@
...
+            )
```
Comment (aolemila) [inline python/sgl_jax/srt/models/kimi_linear.py:313]: According to [huggingface](https://huggingface.co/moonshotai/Kimi-Linear-48B-A3B-Instruct/blob/main/config.json#L56), kimi_linear does not use rope.

You can refer [SGLang's implementation](https://github.com/sgl-project/sglang/blob/c7878dbb6ddfc9c6721b9db20a876f2718b0e955/python/sglang/srt/models/kimi_linear.py#L447).

File: python/sgl_jax/srt/models/kimi_linear.py [hunk 5]
```diff
@@ -0,0 +1,543 @@
...
+                q_lora_rank=config.q_lora_rank,
+                kv_lora_rank=config.kv_lora_rank,
+                qk_nope_head_dim=config.qk_nope_head_dim,
+                qk_rope_head_dim=config.qk_rope_head_dim,
+                v_head_dim=config.v_head_dim,
+                mesh=mesh,
+                layer_id=layer_idx,
+                rope_theta=config.rope_theta,
+                rope_scaling=config.rope_scaling,
+                rope_interleave=True,
```
Comment (aolemila) [inline python/sgl_jax/srt/models/kimi_linear.py]: Remove this parameter because kimi_linear does not support this parameter.

File: python/sgl_jax/srt/models/kimi_linear.py [hunk 6]
```diff
@@ -0,0 +1,543 @@
...
+                    mesh=mesh,
+                    dtype=dtype,
+                )
+            else:
+                self.shared_experts = None
+
+        self.input_layernorm = RMSNorm(
+            config.hidden_size,
+            epsilon=config.rms_norm_eps,
+            param_dtype=jnp.float32,
```
Comment (aolemila) [inline python/sgl_jax/srt/models/kimi_linear.py]: why do you fix the type as fp32?

File: python/sgl_jax/srt/models/kimi_linear.py [hunk 7]
```diff
@@ -0,0 +1,543 @@
...
+        )
+        self.post_attention_layernorm = RMSNorm(
+            config.hidden_size,
+            epsilon=config.rms_norm_eps,
+            param_dtype=jnp.float32,
```
Comment (aolemila) [inline python/sgl_jax/srt/models/kimi_linear.py]: the same to the above

Comment (aolemila) [reply python/sgl_jax/srt/models/kimi_linear.py]: use param_dtype=dtype

File: python/sgl_jax/srt/models/kimi_linear.py [hunk 8]
```diff
@@ -0,0 +1,543 @@
...
+                    dtype=dtype,
+                )
+            else:
+                self.shared_experts = None
+
+        self.input_layernorm = RMSNorm(
+            config.hidden_size,
+            epsilon=config.rms_norm_eps,
+            param_dtype=jnp.float32,
+        )
```
Comment (aolemila) [inline python/sgl_jax/srt/models/kimi_linear.py:402]: add scope_name="input_layernorm"

File: python/sgl_jax/srt/models/kimi_linear.py [hunk 9]
```diff
@@ -0,0 +1,543 @@
...
+        self.post_attention_layernorm = RMSNorm(
+            config.hidden_size,
+            epsilon=config.rms_norm_eps,
+            param_dtype=jnp.float32,
+        )
```
Comment (aolemila) [inline python/sgl_jax/srt/models/kimi_linear.py:408]: add scope_name="post_attention_layernorm"

File: python/sgl_jax/srt/models/kimi_linear.py [hunk 10]
```diff
@@ -0,0 +1,543 @@
...
+                topk=config.num_experts_per_token,
+                renormalize=config.moe_renormalize,
+                num_expert_group=config.num_expert_group,
+                topk_group=config.topk_group,
+                routed_scaling_factor=config.routed_scaling_factor,
+                layer_id=layer_idx,
+            )
+
+            # Expert computation
+            self.block_sparse_moe = EPMoE(
```
Comment (aolemila) [inline python/sgl_jax/srt/models/kimi_linear.py]: Please refer to [qwen3_moe](https://github.com/sgl-project/sglang-jax/blob/c3372a9c040387d4d2f95cf55919ec4074dfbd4f/python/sgl_jax/srt/models/qwen3_moe.py#L203) and keep compatible with `--moe-backend`.

File: python/sgl_jax/srt/models/kimi_linear.py [hunk 11]
```diff
@@ -0,0 +1,543 @@
...
+            epsilon=config.rms_norm_eps,
+            param_dtype=jnp.float32,
+        )
+
+    def __call__(
+        self,
+        positions: jax.Array,
+        hidden_states: jax.Array,
+        forward_batch: ForwardBatch,
+        token_to_kv_pool: KVCache,
```
Comment (aolemila) [inline python/sgl_jax/srt/models/kimi_linear.py]: Please take care of this field. For the full attn layer, this is a token_to_kv_pool. For the linear attention layer, this is a linear recurrent state which includes state and conv_state.

File: python/sgl_jax/srt/models/kimi_linear.py [hunk 12]
```diff
@@ -0,0 +1,543 @@
...
+    ) -> tuple[jax.Array, list, list]:
+        hidden_states = self.embed_tokens(forward_batch.input_ids)
+
+        residual = None
+        layers_kv_fused = []
+        layers_topk_ids = []
+
+        for layer in self.layers:
+            hidden_states, residual, kv_fused, topk_ids = layer(
```
Comment (aolemila) [inline python/sgl_jax/srt/models/kimi_linear.py]: Please check with @Rodrian7  for layer's return.

File: python/sgl_jax/srt/models/kimi_linear.py (hunk 3 above)
Comment (zhengkezhou1) [reply python/sgl_jax/srt/models/kimi_linear.py:310]: done.

File: python/sgl_jax/srt/models/kimi_linear.py (hunk 6 above)
Comment (zhengkezhou1) [reply python/sgl_jax/srt/models/kimi_linear.py]: done.

File: python/sgl_jax/srt/models/kimi_linear.py (hunk 7 above)
Comment (zhengkezhou1) [reply python/sgl_jax/srt/models/kimi_linear.py]: done.

File: python/sgl_jax/srt/models/kimi_linear.py (hunk 8 above)
Comment (zhengkezhou1) [reply python/sgl_jax/srt/models/kimi_linear.py:402]: done.

File: python/sgl_jax/srt/models/kimi_linear.py (hunk 9 above)
Comment (zhengkezhou1) [reply python/sgl_jax/srt/models/kimi_linear.py:408]: done.

File: python/sgl_jax/srt/models/kimi_linear.py (hunk 10 above)
Comment (zhengkezhou1) [reply python/sgl_jax/srt/models/kimi_linear.py]: done.

File: python/sgl_jax/srt/models/kimi_linear.py (hunk 5 above)
Comment (zhengkezhou1) [reply python/sgl_jax/srt/models/kimi_linear.py]: done. removed rope_interleave=True

File: python/sgl_jax/srt/models/kimi_linear.py (hunk 4 above)
Comment (zhengkezhou1) [reply python/sgl_jax/srt/models/kimi_linear.py:313]: done. added `skip_rope` param.

File: python/sgl_jax/srt/models/kimi_linear.py [hunk 13]
```diff
@@ -0,0 +1,754 @@
...
+                    target_path=f"{attn_target}.{proj_name}.weight",
+                    sharding=(None, None),
+                    transpose=True,
+                )
+            mappings[f"{prefix}.self_attn.o_proj.weight"] = WeightMapping(
+                target_path=f"{attn_target}.o_proj.weight",
+                sharding=("tensor", None),
+                transpose=True,
+            )
+            # Conv1d weights: HF shape (projection_size, 1, conv_size) -> JAX (conv_size, projection_size)
```
Comment (zhengkezhou1) [inline python/sgl_jax/srt/models/kimi_linear.py:691]: | Weight | HF Shape | JAX Target Shape |
|--------|----------|-----------------|
| `q/k/v_conv1d.weight` | `(4096, 1, 4)` | `(4, 4096)` |

File: python/sgl_jax/srt/models/kimi_linear.py [hunk 14]
```diff
@@ -0,0 +1,758 @@
...
+            input_size=self.head_dim,
+            output_size=projection_size,
+            mesh=mesh,
+            use_bias=False,
+            params_dtype=dtype,
+            kernel_axes=(None, "tensor"),
+            scope_name="g_b_proj",
+        )
+
+        # activation="sigmoid"
```
Comment (zhengkezhou1) [inline python/sgl_jax/srt/models/kimi_linear.py]: output = o_norm(x) * sigmoid(gate)

File: python/sgl_jax/srt/models/kimi_linear.py [hunk 15]
```diff
@@ -0,0 +1,769 @@
...
+                    dtype=dtype,
+                )
+                for i in range(config.num_hidden_layers)
+            ]
+        )
+
+        self.norm = RMSNorm(
+            config.hidden_size,
+            epsilon=config.rms_norm_eps,
+            param_dtype=jnp.float32,
```
Comment (aolemila) [inline python/sgl_jax/srt/models/kimi_linear.py]: this one

File: python/sgl_jax/srt/models/kimi_linear.py [hunk 16]
```diff
@@ -0,0 +1,769 @@
...
+            if metadata is not None:
+                physical_to_logical_map = np.array(jax.device_get(metadata.physical_to_logical_map))
+                phy_to_log = physical_to_logical_map[layer_idx]
+
+            moe_mappings = create_moe_weights_mapping(
+                prefix=prefix,
+                target_prefix=target_prefix,
+                num_experts=num_logical_experts,
+                expert_type_names=("w1", "w3", "w2"),
+                moe_backend="epmoe",
```
Comment (aolemila) [inline python/sgl_jax/srt/models/kimi_linear.py]: moe_backend=moe_backend

File: python/sgl_jax/srt/models/kimi_linear.py [hunk 17]
```diff
@@ -0,0 +1,769 @@
...
+                )
+                for i in range(config.num_hidden_layers)
+            ]
+        )
+
+        self.norm = RMSNorm(
+            config.hidden_size,
+            epsilon=config.rms_norm_eps,
+            param_dtype=jnp.float32,
+        )
```
Comment (aolemila) [inline python/sgl_jax/srt/models/kimi_linear.py:510]: add scope_name="norm"

File: python/sgl_jax/srt/models/kimi_linear.py (hunk 16 above)
Comment (zhengkezhou1) [reply python/sgl_jax/srt/models/kimi_linear.py]: done.

File: python/sgl_jax/srt/models/kimi_linear.py (hunk 17 above)
Comment (zhengkezhou1) [reply python/sgl_jax/srt/models/kimi_linear.py:510]: done. added scope_name="norm"

**Comment by Prayer3th** [comment]
The fused moe kernel was not considered when the weight was loaded?

**Review by aolemila — APPROVED** [review]

File: python/sgl_jax/test/model/test_kimi_linear.py [hunk 18]
```diff
@@ -0,0 +1,83 @@
+import json
```
Comment (zhengkezhou1) [inline python/sgl_jax/test/model/test_kimi_linear.py]: Currently `python/sgl_jax/test/model/test_kimi_linear.py` is a standalone test file added to CI for real weight loading validation. This doesn't scale — each new model would need its own test file with duplicated boilerplate.

   We should migrate `test_kimi_linear.py` into this framework and add fixtures for other models (DeepseekV3, BailingMoE, etc.) so all model loading tests go through one unified path.
